pkg/config/configfile.go

```diff
@@ -0,0 +1,369 @@
+	if !force {
+		if cf.data != nil {
+			return cf.data, nil
+		}
+	}
+
+	d, e := cf.store.Read(cf.file)
+	if e != nil {
+		return nil, e
+	}
+	cf.data = d
+	return cf.data, nil
+}
+
+// Stat returns the StorageStats for the file.
+func (cf *ConfigFile) Stat() (*storage.StorageInfo, error) {
+	if cf.store == nil {
+		return nil, NoBackingStore
+	}
+
+	return cf.store.Stat(cf.file)
+}
+
+// Exists returns true if the file exist. If an error other than a NotExist error is returned,
+// the result will be false with the provided error.
+func (cf *ConfigFile) Exists() (bool, error) {
+	if cf.store == nil {
+		return false, NoBackingStore
+	}
+
+	return cf.store.Exists(cf.file)
+}
+
+// Delete removes the file from storage permanently.
+func (cf *ConfigFile) Delete() error {
+	if cf.store == nil {
+		return NoBackingStore
+	}
+
+	e := cf.store.Remove(cf.file)
+
+	// on removal, clear data cache
+	if e == nil {
+		cf.dataLock.Lock()
+		cf.data = nil
+		cf.dataLock.Unlock()
+	}
+	return e
+}
+
+// Refresh allows external callers to force reload the config file from internal storage. This is
+// particularly useful when there exist no change listeners on the config, which would prevent the
+// data cache from automatically updating on change
+func (cf *ConfigFile) Refresh() ([]byte, error) {
+	return cf.internalRead(true)
+}
+
+// AddChangeHandler accepts a ConfigChangedHandler function which will be called whenever the implementation
+// detects that a change has been made. A unique HandlerID is returned that can be used to remove the handler
+// if necessary.
+func (cf *ConfigFile) AddChangeHandler(handler ConfigChangedHandler) HandlerID {
+	return cf.AddPriorityChangeHandler(handler, DefaultHandlerPriority)
+}
+
+// AddPriorityChangeHandler allows adding a config change handler with a specific priority. By default,
+// any handlers added via AddChangeHandler have a default priority of 1000. The lower the priority, the
+// sooner in the handler execution it will be called.
+func (cf *ConfigFile) AddPriorityChangeHandler(handler ConfigChangedHandler, priority int) HandlerID {
+	cf.watchLock.Lock()
+	defer cf.watchLock.Unlock()
+
+	h := &pHandler{
+		id:       HandlerID(uuid.NewString()),
+		handler:  handler,
+		priority: priority,
+	}
+
+	cf.watchers = append(cf.watchers, h)
+
+	// create the actual file watcher once we have at least one active watcher func registered
+	if len(cf.watchers) == 1 {
+		cf.runWatcher()
+	}
+
+	return h.id
+}
+
+// RemoveChangeHandler removes the change handler with the provided identifier if it exists. True
+// is returned if the handler was removed (it existed), false otherwise.
+func (cf *ConfigFile) RemoveChangeHandler(id HandlerID) bool {
+	cf.watchLock.Lock()
+	defer cf.watchLock.Unlock()
+
+	for i := range cf.watchers {
+		if cf.watchers[i] != nil && cf.watchers[i].id == id {
+			copy(cf.watchers[i:], cf.watchers[i+1:])
+			cf.watchers[len(cf.watchers)-1] = nil
+			cf.watchers = cf.watchers[:len(cf.watchers)-1]
```

Comment:
Seeing this, I really think `cf.watchers` should be a map, especially because priority sorting happens after copying data out anyway.

Reply:
So this was designed such that equivalent priority should still honor the order in which the handlers were added, and maps do not honor order. Was hoping that would be clear here that I've implemented remove in a way that maintains the existing order. 

It _is_ worth noting that I do not think `sort` makes any guarantees about existing order (as it may rearrange priorities of the same value). If that's the case, then I think a `map` achieves the same goal.

Reply:
Mentioned in Slack, but `slice.SliceStable` (https://pkg.go.dev/sort#SliceStable) is a thing which should give you the behavior you're looking for. If you want to stick with the slice approach because of that, all I request is just a comment on this explaining exactly what you explained to me!

Reply:
For primitive types, this is probably fine. However, for object pointers, the GC can consider an object as referenced if you don't nil out the index and trim the slice appropriately. Append is also slower than copy and I believe it allocates a new slice whereas copy does not.

Reply:
Do you think this code is going to be enough of a performance bottleneck to be worth sacrificing readability for performance?

I would hope that the `append(slice[:s], slice[s+1:]...)` doesn't cause a memory leak, surely once `slice` goes out of scope the objects it references will get GCed?

Reply:
I don't mind if you just want to leave as-is, by the way. I've veered into nit-land here.

Reply:
The slice scope doesn't matter as much as the underlying array the slice points to. `append` allocates a new slice, that slice is super cheap to allocate, but point being that slices are disposable, and even if a 1 length slice points to an element in a  300 length array, that entire array's memory _may_ hang around until _all_ references to that array are gone. 

Directly from the golang github wiki: https://github.com/golang/go/wiki/SliceTricks
![image](https://user-images.githubusercontent.com/334480/144142040-4bd44331-cce8-48ab-934c-cce1f1bbaea4.png)

Reply:
There are more than a few places this code is leveraged, and there is a good reason for it. I don't like the readability argument when it results in leaking memory.

Reply:
Well, that's behavior I wasn't aware of. Good to know and now I'm just a little sad. Thanks for the detail.